# Data

Organized by subfolder.

## `evidence`

Nominally, these files are ultimately used to produce inputs to mCADRE (or MBA, in previous iterations). I've stored them in a separate folder, as they might be useful in their own right. Files are further grouped into subfolders based on source:

### `hpa`

This folder contains tissue- (in this case cell-type) specific protein
expression data, based on antibody staining evidence from the  Human Protein
Atlas (HPA) website. At the time the data were collected (~2010-2012), HPA
didn't have an option to download all information for a specific cell type in
tabular format. Instead, I used some Python scripts to parse HTML tables across
multiple pages and save Ensembl IDs of expressed proteins.

+ **U-87_MG.txt:** Ensembl IDs of proteins expressed in the U-87 cell line,
broken into two lists based on 'Strong' or 'Moderate' evidence (I'd have to go
back and review the code to figure out what my logic was for assigning evidence
levels); this file was generated with the script `getHPA.py`.

+ **U-87_MG_genes.txt:** this should be the same list of genes/proteins as in
**U-87_MG.txt**, but with Ensembl IDs replaced with NCBI Entrez IDs (where
possible); this file was generated with the script `HPAensembl2gene.py` using
the **Homo_sapiens.gene_info** table to look up IDs

+ **U-87_MG_neg.txt:** at some point, HPA added lists/tables of proteins that
stained negatively in cell assays (I collected these lists to potentially use
as a negative control when validating the U87 model); this file was generated
with the script `getHPA_neg.py`

+ **U-87_MG_neg_genes.txt:** same idea as before, with Ensembl IDs converted to
NCBI gene IDs with `HPAensembl2gene.py`

+ **U-138_MG.txt, U-138_MG_genes.txt, U-251_MG.txt:** similar types of HPA data
collected for other glioblastoma cell lines (with the thought to incorporate as
supporting evidence); these were collected pretty early on and never really
used.

### `geo`

This folder contains microarray gene expression data from studies of the U87
cell line, collected from the Gene Expression Omnibus (GEO).

NOTE: For this iteration, I manually downloaded CEL files from GEO and used
some lab code for consensus processing and normalization. For a slightly more organized and reproducible approach, I would recommend referring to the `geo_new` section instead.

+ **all_datasets.mat*:** output of consensus processing/normalization of CEL
files; processing was done with the function `preprocessConsensusPlatform` (which I can't remember came from someone in the Price lab or the Matlab Bioinformatics Toolbox); this file was generated with the script
`U87GEODataProcessing.m` (it is also used as input to the same script, if data
has already been processed)

*this file is pretty large (~40mb); I realized I probably shouldn't put it on
GitHub

+ **U87_GEO_evidence.mat:** data compiled from **all_datasets** .mat file
described above and formatted for input to MBA/mCADRE; this file was generated by the script 'U87GEODataProcessing.m'; it is used as input to the script
`createU87mCADREInputs.m`

+ **U87_GEO_10272013.mat:** newer and slightly different version of
**U87_GEO_evidence.mat**; can't tell how it was generated or whether it was
used anywhere (should probably delete)

+ **U-87_MG_GEO.txt:** I believe this is just a union list of all genes from
the compiled GEO data.. or it might be those with non-zero expression - I'm not
entirely sure.

+ **U-87_MG_GEO_genes.txt:** however **U-87_MG_GEO.txt** was generated and
whatever it represents, gene symbols were converted to NCBI gene IDs using the
script `GEOsymbol2gene.py`.

### `geo_new`

This folder includes all files from a more recent effort to collect U87
microarray data from GEO, this time using a somewhat more systematic and
documented approach.

+ **GEO Datasets_GBM.xlsx:** this table describes the GEO datasets (GSE
accession ID) identified for U87, including what information I could find about
media conditions used in the study (note: I tried to limit collection to
datasets generated on Affymetrix platforms); the table also includes a list of
additional datasets that might also be worth incorporating; this table is
well over *two years* old, so it should definitely not be considered
exhaustive.

+ **sampleList.txt:** for the seven GEO datasets listed in the first table of
**GEO Datasets_GBM.xlsx**, this file provides the ID (GSM accession) of the
sample most relevant to U87 as well as the platform ID for the dataset; this
file is used by the script `importU87AffyData.R` to download data from GEO.

+ **GSE[x].txt:** MAS5 presence/absence calls for probes in each dataset in
**sampleList.txt**; these files were generated by `importU87AffyData.R`.

+ **GSE[x]_map.txt:** for each probe in the corresponding MAS5 data above,
these files include the Entrez gene ID (from `biomaRt`); these files were also
generated by `importU87AffyData.R`.

+ **U87_GEO_Jan2014.mat:** this file includes GEO data, combined from the
above datasets and formatted for input to mCADRE (i.e., ubiquity scores); this
file was generated by the script `compileGEOData.m`.

### `lab`



## `simulation`

This folder includes relevant information for defining parameters and constraints in model simulations (whether for testing, pre-processing, model-building, or validation). Files are further grouped into subfolders based on type of information:

### `biomass`

### `media`

### `experiments`


## `models`

This folder contains different versions of both the generic (Human Recon) and U87-specific COBRA models, along with some related metadata. Files are grouped into subfolders based on generic or cell-type specific model:

### `recon2`

### `u87`


## `mcadre`

This folder contains the final inputs for and outputs from mCADRE.


## `reference`

This folder contains some reference data from NCBI, mostly for tasks that could be done much more easily with `biomaRt`...


## other

+ **`Workbook3.xlsx`:** I have no idea what this file is, but I'm going to hang onto it until I'm confident it's not important.
